core/src/main/java/org/apache/druid/java/util/common/guava/ParallelMergeCombiningSequence.java

```diff
@@ -0,0 +1,1071 @@
+    {
+      try {
+        long start = System.nanoTime();
+
+        int counter = 0;
+        int batchCounter = 0;
+        ResultBatch<T> outputBatch = new ResultBatch<>(batchSize);
+
+        T currentCombinedValue = initialValue;
+        while (counter++ < yieldAfter && !pQueue.isEmpty()) {
+          BatchedResultsCursor<T> cursor = pQueue.poll();
+
+          // push the queue along
+          if (!cursor.isDone()) {
+            T nextValueToAccumulate = cursor.get();
+
+            cursor.advance();
+            if (!cursor.isDone()) {
+              pQueue.offer(cursor);
+            } else {
+              cursor.close();
+            }
+
+            // if current value is null, combine null with next value
+            if (currentCombinedValue == null) {
+              currentCombinedValue = combineFn.apply(null, nextValueToAccumulate);
+              continue;
+            }
+
+            // if current value is "same" as next value, combine them
+            if (orderingFn.compare(currentCombinedValue, nextValueToAccumulate) == 0) {
+              currentCombinedValue = combineFn.apply(currentCombinedValue, nextValueToAccumulate);
+              continue;
+            }
+
+            // else, push accumulated value to the queue, accumulate again with next value as initial
+            outputBatch.add(currentCombinedValue);
+            batchCounter++;
+            if (batchCounter >= batchSize) {
+              outputQueue.offer(outputBatch);
+              outputBatch = new ResultBatch<>(batchSize);
+              batchCounter = 0;
+            }
+
+            // next value is now current value
+            currentCombinedValue = combineFn.apply(null, nextValueToAccumulate);
+          } else {
+            cursor.close();
+          }
+        }
+
+        if (!pQueue.isEmpty() && !cancellationGizmo.isCancelled()) {
+          // if there is still work to be done, execute a new task with the current accumulated value to continue
+          // combining where we left off
+          if (!outputBatch.isDrained()) {
+            outputQueue.offer(outputBatch);
+          }
+
+          // measure the time it took to process 'yieldAfter' elements in order to project a next 'yieldAfter' value
+          // which we want to target a 10ms task run time. smooth this value with a cumulative moving average in order
+          // to prevent normal jitter in processing time from skewing the next yield value too far in any direction
+          final long elapsedMillis = Math.max(
+              TimeUnit.MILLISECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS),
+              1L
+          );
+          final double nextYieldAfter = Math.max(10.0 * ((double) yieldAfter / elapsedMillis), 1.0);
+          final double cumulativeMovingAverage = (nextYieldAfter + (recursionDepth * yieldAfter)) / (recursionDepth + 1);
+          final int adjustedNextYieldAfter = (int) Math.ceil(cumulativeMovingAverage);
```

Comment:
assuming we really want task to preempt itself in about 10ms, I wonder why we couldn't just check at the `elasped_time_so_far` inside the loop whenever one batch was pushed i.e. inside the `if (batchCounter >= batchSize) {` block at line-495 .  if  `elasped_time_so_far` was greater than `10 ms` then break from the loop at that point? that would probably more accurately evict the task in about `10 ms`

also, it would be nice to know your mental notes on why you chose threshold to be `10 ms` .

Reply:
maybe it would've been too expensive to do on each batch completion ... but in that case it could be checked on every, say 10 batch completion.

not saying that this necessarily needs to change but trying to understand how the choice is made.

Reply:
The 10ms is from @leventov who came up with this approach in general in [a thread in a previous attempt at parallel broker merges](https://github.com/apache/incubator-druid/pull/6629#discussion_r241089247), and I rolled with it because it seemed like a reasonable time period to get a decent amount of work done without starving other queries from running their tasks on the pool. Disclaimer that I haven't done a lot of measuring with other task running time targets since this one has been benchmarking reasonably well.

I haven't measure the difference, but I don't think it needs to be incredibly strict, was just shooting for close enough within reason, so it seemed cheaper to check the actual elapsed time less frequently. But agree it might be worth investigating, if I find some time I'll try to play around with it.

Reply:
Idea is pre-emption is to give way to other work submitted to the FJP, so it should take `fjp.getQueuedSubmissionCount()` into consideration. If `fjp.getQueuedSubmissionCount() = 0` , that means there is no work pending on cpu availability, so there is no point in giving up the cpu, it should continue going.

Now, for the other case when there is queued work waiting for CPU, then we definitely wanna pre-empt, now the choice of 10ms vs 100ms is arbitrary and based purely on gut feeling. I would've chosen it to be 100ms just to err on the side of making same combine operation more efficient and not hogging CPU for a ridiculously long time.

Now, when we talk about 10ms above .. we want to measure the amount of time combine op got actually being on the CPU. But, in the while loop I see, there are many places where we are making potentially blocking operations .. so 
```
          final long elapsedMillis = Math.max(
              TimeUnit.MILLISECONDS.convert(System.nanoTime() - start, TimeUnit.NANOSECONDS),
              1L
          );
```
is giving us potentially way more than the time spent on CPU and `ThreadMXBean.getCurrentThreadCpuTime()` could be used to get that more accurately instead of `System.nanoTime()` .  (I am assuming that call to `ForkJoinPool.managedBlock(..)` doesn't change thread but comparing currentThreadID at start and end, and printing a warning if threadID changed would be good to know if that assumption is incorrect).

does it make sense ?

Reply:
>Idea of pre-emption is to give way to other work submitted to the FJP, so it should take fjp.getQueuedSubmissionCount()

My worry about using this and other fjp inspection methods too frequently was that they look semi expensive, looping over the work queues to count stuff, but I can look into it a bit more and do some experiments, as maybe measurement overhead is insignificant.

>The choice of 10ms vs 100ms is arbitrary and based purely on gut feeling.

Yes this is true, I think I could likely make it configurable to assist in tuning.

>Now, when we talk about 10ms above .. we want to measure the amount of time combine op got actually being on the CPU. But, in the while loop I see, there are many places where we are making potentially blocking operations 

Yeah, I guess I was more approaching it as time slicing the thread pool to try to minimize the number of extra threads the pool needs to create for normal operation when stuff is blocking, rather than directly considering the cpu usage, but what you are suggesting might make more sense because that is measuring the actual work it is doing, I don't have a strong gut feeling on either way so did the easiest approach to measurement. 

At least the way stuff is currently structured, I don't believe I can currently fully take the potentially blocking parts out of the time measurement unless I do a lot more, smaller measurements and combine them, measuring only periods where I know blocking will not happen. 

I will do some more experiments and see what I can do to determine what is the most effective way to do the measurements, what actually should be measured, as well as how they should be used in terms of task run time (though easier said than done I think).

Reply:
I know giving ideas is easy and  totally agree about the "easier said than done" part , and you have the first right to any choice or to degree of experimentations done as the author :)

Looked at `fjp.getQueuedSubmissionCount()` and indeed it is fairly expensive at busy times, OTOH noticed that there is a `fjp.hasQueuedSubmissions()` which is much faster and just tells if `fjp.getQueuedSubmissionCount() > 0` 

> At least the way stuff is currently structured, I don't believe I can currently fully take the potentially blocking parts out of the time measurement unless I do a lot more, smaller measurements and combine them, measuring only periods where I know blocking will not happen.

yeah, Even if you did that , cpu time is more effective because so called "blocking" code in `ManagedBlocker` isn't necessarily blocking from OS perspective every time and could consume cpu time.

Reply:
Regardless of what value is chosen (or a config parameter introduced), there should be an extensive comment describing the default choice. Currently, there is a magic constant.

[Make the code obvious by adding **a lot** of explaining comments](https://github.com/apache/incubator-druid/blob/fe5bb66aad5ba3ce808a7bf243a6ef24240d7507/dev/principles.md#explaining-comments).

Reply:
>I am assuming that call to ForkJoinPool.managedBlock(..) doesn't change thread but comparing currentThreadID at start and end, and printing a warning if threadID changed would be good to know if that assumption is incorrect

I dug around deeper into the code for `ForkJoinPool` and for good measure added the `log.warn` and did some tests, and thread id does appear stable through a managed block, which makes sense to me. From my understanding, work does not move to a new thread, it is just that the pool just no longer counts the thread as 'active', allowing the pool to spawn a new thread which presumably can actually do work while this one chills out. The blocked thread becomes active again, as far as the pool is concerned, once the managed blocker becomes release-able. Then the pool will try to shrink the overall thread count back down as it is able.

Reply:
Thanks for digging in on this @himanshug and making me really have a closer look at this area in particular, since I think it really is one of the most important parts to ensure that the pool operates well when it is heavily saturated. I've been working on a benchmark that directly tests `ParallelMergeCombiningSequence` in order to try out the differences between 10ms and 100ms, using wall time vs using cpu time, behavior when overloaded, etc. 

While I haven't quite got to that yet, in the process I have discovered a mistake that heavily impacted performance when the pool was overloaded when computing the level of parallelism: https://github.com/apache/incubator-druid/pull/8578/files#diff-84792f9d3cefe47cbb471669dce2a276R374, specifically that it was using the pool parallelism instead of the processor count, which meant it was overestimating it's capacity for work. These benchmarks are not pushed yet, and will probably change a lot as I clean them up, but the rough explanation is

* `testJustOne` uses `ParallelMergeCombiningSequence` and a single jmh thread on empty fork join pool
* `testManyParallel` does 24 jmh threads with `ParallelMergeCombiningSequence` all being evaluated at the same time (this was my suspected worst case for the greedy parallelism chooser)
* `testManyParallelJitter` does 24 jmh threads with a random 0-50ms delay before starting work on the `ParallelMergeCombiningSequence` to give a more realistic load pattern where not everything is starting at exactly the same time
* `testSerial` uses `CombiningSequence` wrapping a `MergeSequence` to do equivalent operation to `ParallelMergeCombiningSequence` 
* `testSerialParallel` is 24 threads of `testSerial`
* `testSerialParallelJitter` is `testSerialParallel` with 0-50ms of random jitter

Benchmarks were run twice both with and without synchronized benchmark iterations. Synchronized to show performance of the exact scenario described for each benchmark above, and without synchronization which lets each thread do their iterations at their leisure and more likely emulating an actual sustained heavy load pattern. 

```
pool parallelism - running + queued
(no sync)
Benchmark                                                         (batchSize)  (numSequences)  (parallelism)  (rowsPerSequence)  (targetTaskTimeMillis)  (yieldAfter)  Mode  Cnt     Score     Error  Units
ParallelMergeCombiningSequenceBenchmark.testJustOne                       128               8              4              75000                      10          1024  avgt   20    60.393 ±   4.754  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallel                  128               8              4              75000                      10          1024  avgt   20  1389.137 ± 904.866  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallelJitter            128               8              4              75000                      10          1024  avgt   20   647.165 ± 168.406  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerial                        128               8              4              75000                      10          1024  avgt   20    75.212 ±   0.600  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallel                128               8              4              75000                      10          1024  avgt   20   509.219 ±   9.345  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallelJitter          128               8              4              75000                      10          1024  avgt   20   505.975 ±  12.986  ms/op

(sync)
Benchmark                                                         (batchSize)  (numSequences)  (parallelism)  (rowsPerSequence)  (targetTaskTimeMillis)  (yieldAfter)  Mode  Cnt     Score      Error  Units
ParallelMergeCombiningSequenceBenchmark.testJustOne                       128               8              4              75000                      10          1024  avgt   20    61.379 ±    5.957  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallel                  128               8              4              75000                      10          1024  avgt   20  1883.554 ± 1614.321  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallelJitter            128               8              4              75000                      10          1024  avgt   20   540.248 ±   17.194  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerial                        128               8              4              75000                      10          1024  avgt   20    76.956 ±    0.768  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallel                128               8              4              75000                      10          1024  avgt   20   515.362 ±   14.206  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallelJitter          128               8              4              75000                      10          1024  avgt   20   507.267 ±   10.427  ms/op
```

after the change to `final int computedParallelism = maxParallelism - utilizationEstimate`:

```
num cores - (running + queue)
(no sync)
Benchmark                                                         (batchSize)  (numSequences)  (parallelism)  (rowsPerSequence)  (targetTaskTimeMillis)  (yieldAfter)  Mode  Cnt    Score    Error  Units
ParallelMergeCombiningSequenceBenchmark.testJustOne                       128               8              4              75000                      10          1024  avgt   20   49.720 ±  1.188  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallel                  128               8              4              75000                      10          1024  avgt   20  533.131 ± 14.251  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallelJitter            128               8              4              75000                      10          1024  avgt   20  539.346 ± 13.348  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerial                        128               8              4              75000                      10          1024  avgt   20   75.876 ±  0.579  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallel                128               8              4              75000                      10          1024  avgt   20  500.047 ± 12.856  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallelJitter          128               8              4              75000                      10          1024  avgt   20  517.185 ± 11.089  ms/op

(sync)
Benchmark                                                         (batchSize)  (numSequences)  (parallelism)  (rowsPerSequence)  (targetTaskTimeMillis)  (yieldAfter)  Mode  Cnt    Score    Error  Units
ParallelMergeCombiningSequenceBenchmark.testJustOne                       128               8              4              75000                      10          1024  avgt   20   49.861 ±  1.088  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallel                  128               8              4              75000                      10          1024  avgt   20  522.590 ± 16.520  ms/op
ParallelMergeCombiningSequenceBenchmark.testManyParallelJitter            128               8              4              75000                      10          1024  avgt   20  530.802 ± 18.761  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerial                        128               8              4              75000                      10          1024  avgt   20   75.877 ±  0.882  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallel                128               8              4              75000                      10          1024  avgt   20  517.718 ± 12.978  ms/op
ParallelMergeCombiningSequenceBenchmark.testSerialParallelJitter          128               8              4              75000                      10          1024  avgt   20  520.770 ±  9.040  ms/op
```

After the modification, things look a lot better. I'm going to clean up the benchmarks and probably split out the same threaded combining sequence wrapping a merge sequence reference benchmarks so they don't get run with parameters that don't apply, which will allow me to dig deep into the tuning matters about target task run time and how to measure it, batch size, as well as some additional variations like sequences with occasional blocking periods and stuff.